### PR TITLE
refactor: change the way action filter checks custom attribute in controller

### PR DESCRIPTION
**Pull request by alshuriga**
I've changed the way of mvc action filter checking if controller has FluentValidationAutoValidation attribute.
Tested it in WeatherForecast (dotnet webapi template app) and it seems to work fine.

### CODE REVIEW

File: .vs/FluentValidation.AutoValidation/FileContentIndex/0b76e120-00f5-4fdb-8a89-982690eeb349.vsidx
Comment (mvdgun) [inline .vs/FluentValidation.AutoValidation/FileContentIndex/0b76e120-00f5-4fdb-8a89-982690eeb349.vsidx:1]: remove

File: .vs/FluentValidation.AutoValidation/FileContentIndex/read.lock
Comment (mvdgun) [inline .vs/FluentValidation.AutoValidation/FileContentIndex/read.lock:1]: remove

File: .vs/FluentValidation.AutoValidation/v17/TestStore/0/000.testlog
Comment (mvdgun) [inline .vs/FluentValidation.AutoValidation/v17/TestStore/0/000.testlog:1]: remove

File: .vs/FluentValidation.AutoValidation/v17/TestStore/0/testlog.manifest
Comment (mvdgun) [inline .vs/FluentValidation.AutoValidation/v17/TestStore/0/testlog.manifest:1]: remove

**Review by mvdgun — CHANGES_REQUESTED** [review]
@alshuriga thanks for your contribution! I am seeing some binary files in your PR, perhaps some artifacts from your IDE? Could you remove those? I've marked those files for you.

**Comment by mvdgun** [comment]
Looking back at the code I think the annotations on methods will break since it only retrieves annotations on the type. Did you test putting an annotation on a controller method?

**Comment by alshuriga** [comment]
> Looking back at the code I think the annotations on methods will break since it only retrieves annotations on the type. Did you test putting an annotation on a controller method?

thank you fot noticing that, it works now for both controller and action annotation.

**Comment by mvdgun** [comment]
@alshuriga I'm still debating whether this change will have a detrimental impact on performance since it uses reflection. Where possible I would like to avoid using reflection. I'm going to look into the `EndpointMetadata` some more. Maybe .NET offers some public API's that expose the applied attributes.

**Comment by alshuriga** [comment]
@mvdgun Yes, GetCustomAttribute actually does impact performance if called every request. I found another way to retrieve attributes, it actually seems to work very similar to the original logic, but there's no restrictions in the documentation for using it in application code. Please take a look.

File: FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs [hunk 1]
```diff
@@ -11,6 +13,7 @@
 using SharpGrip.FluentValidation.AutoValidation.Mvc.Enums;
 using SharpGrip.FluentValidation.AutoValidation.Shared.Extensions;
 
+
```
Comment (mvdgun) [inline FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs:16]: @alshuriga remove empty line

File: FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs [hunk 2]
```diff
@@ -2,7 +2,9 @@
 using System.Linq;
 using System.Threading.Tasks;
 using FluentValidation;
+using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc;
+
```
Comment (mvdgun) [inline FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs:7]: @alshuriga remove empty line

File: FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs [hunk 3]
```diff
@@ -30,8 +33,7 @@ public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionE
             {
                 var actionDescriptor = context.ActionDescriptor;
 
-                // @todo figure out a better way to retrieve the attribute since using the `context.ActionDescriptor.EndpointMetadata` is not recommended for application code
-                if (autoValidationMvcConfiguration.ValidationStrategy == ValidationStrategy.Annotations && !actionDescriptor.EndpointMetadata.OfType<FluentValidationAutoValidationAttribute>().Any())
+                if (autoValidationMvcConfiguration.ValidationStrategy == ValidationStrategy.Annotations && !context.HttpContext.GetEndpoint().Metadata.OfType<FluentValidationAutoValidationAttribute>().Any())
```
Comment (mvdgun) [inline FluentValidation.AutoValidation.Mvc/src/Filters/FluentValidationAutoValidationActionFilter.cs:36]: @alshuriga the `HttpContext.GetEndpoint()` method can return null. Can you move the endpoint to a new variable called `endpoint` above the `var actionDescriptor` and check for null in the if at line 36?

**Review by mvdgun — CHANGES_REQUESTED** [review]
@alshuriga Looks like a great option! I have added some remarks.

**Comment by mvdgun** [comment]
Closes #4

**Review by mvdgun — APPROVED** [review]

**Comment by mvdgun** [comment]
Merged, thanks for you contribution @alshuriga!

**Pull request merged**